File: frontend/src/Component/Todo.js
import React, { useEffect, useState } from 'react';
import Task from './Task';
import InputBox from './InputBox';
import withHover from './Hoverable';
import Header from './Header';
import PostAPI from './PostAPI';
import '../ComponentCss/Todo.css';

const Todo = () => {
  const [todo, setTodo] = useState({ heading: '', tasks: [] });

  const updateTodo = () => {
    PostAPI.getTodo().then(setTodo);
  };

  useEffect(updateTodo, []);

  const updateStatus = (taskId) => {
    PostAPI.updateStatus(taskId).then(updateTodo);
  };

  const addTask = (content) => {
    PostAPI.addTask(content).then(updateTodo);
  };

  const deleteTask = (taskId) => {
    PostAPI.deleteTask(taskId).then(updateTodo);
  };

  const updateHeading = (title) => {
    PostAPI.updateHeading(title).then(updateTodo);
  };

  const resetTodo = () => {
    PostAPI.resetTodo().then(updateTodo);
  };

  const tasks = todo.tasks.map((task, index) => {
    const HoverableTask = withHover(Task, deleteTask);

    return (
      <HoverableTask
        task={task}
        taskId={index}
        key={index}
        onClick={updateStatus}
      />
    );
  });

  const HoverableHeader = withHover(Header, resetTodo);
  return (
    <div className="todo">
      <HoverableHeader heading={todo.heading} onChange={updateHeading} />
      <div> {tasks}</div>
      <InputBox onChange={addTask} />
    </div>
  );
};

export default Todo;

File: backend/src/main/java/com/todo/services/Todo.java
package com.todo.services;

import java.util.ArrayList;
import java.util.List;

public class Todo{

    private final List<Task> tasks;
    private String heading;


    public Todo(){
        this.tasks = new ArrayList<>();
        this.heading = "Todo";
    }

    public List<Task> getTasks() {
        return tasks;
    }

    public String getHeading() {
        return heading;
    }

    public void addTask(String task) {
        tasks.add(new Task(task, 0));
    }

    public void updateStatus(int taskId) {
        Task task = tasks.get(taskId);
        task.updateStatus();
    }

    public void deleteTask(int taskId) {
        tasks.remove(taskId);
    }

    public void updateHeading(String heading) {
        this.heading =  heading;
    }

    public void resetTodo() {
        this.tasks.clear();
        this.heading = "Todo";
    }
}

